### PR TITLE
Map Degrader

### PR DESCRIPTION
**Pull request by Zemurin**
It degrades map resolution and detail.

### CODE REVIEW

**Comment by Zemurin** [comment]
need appveyor/travis support.

File: MapDegrader/Source/Mappers/LandedTitlesScraper/LandedTitles.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#include "Parser.h"
+
+class LandedTitles: commonItems::parser
+{
+public:
+	void loadTitles(std::istream& theStream, const std::string& theDJLiege, const std::string& theName);
+	void loadTitles(const std::string& fileName);
+	void loadTitles(std::istream& theStream);
+
+	[[nodiscard]] const auto& getTitles() const { return titles; }
```
Comment (Idhrendur) [inline MapDegrader/Source/Mappers/LandedTitlesScraper/LandedTitles.h]: Mild preference for these being in the same order as the members.

File: MapDegraderTests/MapperTests/ColorMapTests.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	definitions.loadDefinitions(definitionStream);
+
+	ColorMapper colorMapper;
+	colorMapper.craftReplacementColorMatrix(titles, definitions);
+
+	const auto& matrix = colorMapper.getReplacementMatrix(); // matrix tells us which colors to replace with which others.
+
+	ASSERT_EQ(2, matrix.size());
+	
+	const auto& pair1 = matrix[0];
```
Comment (Idhrendur) [inline MapDegraderTests/MapperTests/ColorMapTests.cpp]: You should be able to use structured for better naming:
```const auto& [oldColor1, newColor1] = matrix[0];```

File: MapDegraderTests/MapperTests/ColorMapTests.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	const auto& pair2 = matrix[1];
```
Comment (Idhrendur) [inline MapDegraderTests/MapperTests/ColorMapTests.cpp]: Ditto
```const auto& [oldColor2, newColor2] = matrix[1];```

File: MapDegraderTests/MapperTests/ColorMapTests.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	colorMapper.craftReplacementColorMatrix(titles, definitions);
+
+	std::stringstream locStream;
+	locStream << "l_english:\n";
+	locStream << " c_county3:0 \"The County\" # comment\n";
+	LocalizationScraper localizations;
+	localizations.loadLocalizations(locStream);
+
+	// This is expected result. County3 mapped to highest barony # (17), using that barony's colors.
+	const auto* output = "ProvID;r;g;b;title;x;\n17;7;8;9;The County;x;\n";
```
Comment (Idhrendur) [inline MapDegraderTests/MapperTests/ColorMapTests.cpp]: Please split string by lines for better readability:
```
const auto* output = "ProvID;r;g;b;title;x;\n"
    "17;7;8;9;The County;x;\n";
```

File: MapDegraderTests/MapperTests/ColorMapTests.cpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+	ColorMapper colorMapper;
+	colorMapper.craftReplacementColorMatrix(titles, definitions);
+
+	std::stringstream locStream;
+	LocalizationScraper localizations;
+	localizations.loadLocalizations(locStream);
+
+	// Locs will default to title name
+	const auto *output = "ProvID;r;g;b;title;x;\n17;7;8;9;c_county3;x;\n";
```
Comment (Idhrendur) [inline MapDegraderTests/MapperTests/ColorMapTests.cpp]: Ditto

File: MapDegrader/Source/Mappers/ColorMapper/ColorMapper.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+{
+public:
+	void craftReplacementColorMatrix(const LandedTitles& landedTitles, const Definitions& definitions);
+	void exportDefinitions(const LocalizationScraper& localization) const;
+	[[nodiscard]] std::string exportDefinitionsToString(const LocalizationScraper& localization) const;
+
+	[[nodiscard]] const auto& getReplacementMatrix() const { return replacementMatrix; }
+
+private:
+	std::map<std::string, std::map<int, commonItems::Color>> countyBaronies; // we ping off the LAST barony in map (largest key).
```
Comment (Idhrendur) [inline MapDegrader/Source/Mappers/ColorMapper/ColorMapper.h]: This can be local to craftReplacementColorMatrix()

**Review by Idhrendur — APPROVED** [review]

**Pull request merged**